apps/snitch_core/lib/core/domain/shipment_engine.ex

```diff
@@ -0,0 +1,152 @@
+defmodule Snitch.Domain.ShipmentEngine do
+  @moduledoc """
+  Implements functions to handle creation of the optimal shipment.
+
+  > Note 
+    No support for partially fulfilled and backordered packages.
+  """
+
+  import Aruspex.Problem, only: [new: 0, add_variable: 3, post: 4, post: 3]
+  alias Snitch.Data.Schema.Order
+  alias Aruspex.Strategy.SimulatedAnnealing
+
+  @domain [true, false]
+
+  @doc """
+  Returns a list of packages fulfilling an order
+
+  Takes a list of `packages` and an `order` struct.
+  """
+  @spec run(list, Order.t()) :: list
+  def run(packages, order) do
+    edges = create_csp(packages)
+    pckgs = append_keys(packages)
+
+    vars =
+      Enum.reduce(pckgs, [], fn p, acc ->
+        [p.id | acc]
+      end)
+
+    item_var_map = item_var_mapping(pckgs)
+
+    problem = new()
+    domain = @domain
+
+    for var <- vars, do: add_variable(problem, var, domain)
+
+    binary_constraint(problem, edges)
+    summation_constraint(problem, vars, item_var_map, order)
+
+    problem
+    |> SimulatedAnnealing.set_strategy()
+    |> Enum.take(1)
+    |> case do
+      [] ->
+        []
+
+      [result] ->
+        bindings = variable_assignment(result)
+        filter_packages(item_var_map, bindings)
+    end
+  end
+
+  defp variable_assignment(result) do
+    case result do
+      [] ->
+        []
+
+      data ->
+        Enum.reject(data.binding, fn
+          {{:hidden, _, _}, _} -> true
+          _ -> false
+        end)
+    end
+  end
+
+  defp filter_packages(pckgs, bindings) do
+    Enum.reduce(bindings, [], fn
+      {id, true}, acc -> [pckgs[id] | acc]
+      {_, false}, acc -> acc
+    end)
+  end
+
+  defp binary_constraint(problem, edges) do
+    Enum.map(edges, fn {x, y} ->
+      post(problem, x.id, y.id, &(not (&1 and &2)))
+    end)
+  end
+
+  defp summation_constraint(problem, vars, item_var_map, order) do
+    post(problem, vars, fn values ->
+      item_count =
+        vars
+        |> Enum.zip(values)
+        |> Enum.reduce(0, fn
+          {var, true}, acc -> length(item_var_map[var].items) + acc
+          {_, false}, acc -> acc
+        end)
+
+      item_count == length(order.line_items)
+    end)
+  end
+
+  defp item_var_mapping(packages) do
+    Enum.map(packages, fn pkg ->
+      {pkg.id, pkg}
+    end)
+  end
+
+  defp create_csp(packages) do
+    packages
+    |> append_keys()
+    |> Enum.map(&group_package_with_item(&1))
+    |> List.flatten()
+    |> group_packages_by_item()
+    |> packages_with_same_item()
+    |> filter_self_edges()
+    |> find_unique_edges()
+  end
+
+  defp append_keys(packages) do
```

Comment:
You could have used [`Stream.with_index/2`](https://hexdocs.pm/elixir/Stream.html#with_index/2).

Reply:
`Stream.with_index` need indexes to be an integer, here atoms are being used for identification.